mgr/dashboard: introduce gather facts in host list

avanthakkar opened the pull request:
Introducing hosts facts in Dashboard; modifying the host list endpoint with a facts boolean query parameter. Also introducing facts in expand cluster review page.

**If Orchestrator available:**
![Screenshot from 2021-09-14 14-29-21](https://user-images.githubusercontent.com/23611106/133228220-93000359-cd1a-44f8-9ded-a0ea77ef4a46.png)

**If Orchestrator is not available:**
![Screenshot from 2021-09-15 15-57-02](https://user-images.githubusercontent.com/23611106/133417492-11c667ae-2840-412f-a39a-8c93ed854c9e.png)

**Expand cluster updated review page  :**
![Screenshot from 2021-09-30 19-44-09](https://user-images.githubusercontent.com/23611106/135472198-6e10cfe1-8974-4d62-ae8f-e338b11cac4c.png)



**Total Memory**: total_memory
**Raw capacity**: Total actual storage capacity
**Disk Types**: different disks hdd, ssd, nvme and their counts
**Link Speed(NIC)**: speed of Ethernet  interface

Fixes: https://tracker.ceph.com/issues/52017, https://tracker.ceph.com/issues/51848
Signed-off-by: Avan Thakkar <athakkar@redhat.com>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

epuertat commented:
Looks good @avanthakkar. Here it goes:
* It's a total of 11 columns... I don't see an easy way to reduce that to a more reasonable number (the magic 7), but perhaps we can simplify some things:
  * The services row, I would just print the first part of each service (the service type: `mon`, `mgr`...) and hide the common hostanme (`.ceph-node-00.cephlab.com`). We can display the whole thing on mouseover/tooltip.
  * Same for the version: let's just print the major numbers (`17.0.0`) and only display the whole (`17.0.0-xxxx-zzzzzzzz`) on mouse over. In fact, if what we want to highlight is a mismatching version, we should rather find other ways of doing it (coloring?).
* Memory: as the facts gathered are not refreshed that often, the available memory might be out of sync, so perhaps we should only display the total one.
* Disk types is '4'? Maybe disk number?
* NIC: in upper case as it's an acronym.

avanthakkar commented:
### Update 
Following this design https://marvelapp.com/prototype/5ed82bc/screen/73191926

**CPUs:** cpu_count (I'm not sure if this info is relevant to show in table. Thoughts @pcuzner @epuertat ?)

**Raw Capacity:**  `hdd_capacity + flash_capacity ` .  I'm not much familiar with `flash_capacity` field in gather_facts (does it includes `SSD` & `NVMes`) , @pcuzner  maybe you familiar with it?

**Disk Types:**  disk types with their count. How to show counts for SSD/NVME , maybe to use `flash_count` field from gather_facts?

**NIC:** I think in the design it represents the bandwidth of a Ethernet interface maybe, but I don't see any corresponding info in gather_facts, it just exposes the `nic_count`.

I'm attaching the [payload](https://pastebin.com/rZq421vt) of gather_facts too, in case if any other relevant fields needs to be exposed in host list.

avanthakkar commented:
> Looks good @avanthakkar. Here it goes:
> 
>     * It's a total of 11 columns... I don't see an easy way to reduce that to a more reasonable number (the magic 7), but perhaps we can simplify some things:
>       
>       * The services row, I would just print the first part of each service (the service type: `mon`, `mgr`...) and hide the common hostanme (`.ceph-node-00.cephlab.com`). We can display the whole thing on mouseover/tooltip.
I'm not sure if we can do this, as it's linked with perf_counters for now, so let's say some host has **multiple** osd daemons and we just display `osd ` in services column, then how we redirect to perf_counters, as it exists per `service_id` in this case `/osd.1` , `osd.2` and so on. So for now I'll leave it as it is and maybe in future we can expose endpoint by `service_types` too. I see 1 [tracker](https://tracker.ceph.com/issues/46236) for per_counter endpoint improv.
>       * Same for the version: let's just print the major numbers (`17.0.0`) and only display the whole (`17.0.0-xxxx-zzzzzzzz`) on mouse over. In fact, if what we want to highlight is a mismatching version, we should rather find other ways of doing it (coloring?).
> 
I think user expects to see the same version as CLI cmd ..maybe they don't notice mouse hover for it. Mismatching version with what? I didn't get it.
>     * Memory: as the facts gathered are not refreshed that often, the available memory might be out of sync, so perhaps we should only display the total one.
> 
Yes I agree. I have updated the screenshot too accordingly.
>     * Disk types is '4'? Maybe disk number?
> 
>     * NIC: in upper case as it's an acronym.

pereman2 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:
```diff
@@ -184,6 +186,37 @@ export class HostsComponent extends ListWithDetails implements OnInit {
+        flexGrow: 1
+      },
+      {
+        name: $localize`Disk Types`,
```
As Ernesto said, I still think this should be `Disk Count`, `Disk Number` or something like that. With `Disk Types` I would expect a list of types like the labels column. What do you think @avanthakkar?

avanthakkar replied:
> As Ernesto said, I still think this should be `Disk Count`, `Disk Number` or something like that. With `Disk Types` I would expect a list of types like the labels column. What do you think @avanthakkar?

Yes there's disk types two..if u see the screenshot attached in PR description it's not just count, it says `4 HDD` so it also represents type of disk.

pereman2 replied:
I meant that with `types` I wouldn't expect to see a number but with `number` seeing `hdd` is ok. Probably I'm being super picky, `Disk Types` works fine too.

avanthakkar replied:
I changed it to `Disks` now to cover both types and count..what u think? @pereman2

epuertat commented:
> ### Update
> Following this design [marvelapp.com/prototype/5ed82bc/screen/73191926](https://marvelapp.com/prototype/5ed82bc/screen/73191926)
> 
> **CPUs:** cpu_count (I'm not sure if this info is relevant to show in table. Thoughts @pcuzner @epuertat ?)
> 
> **Total Memory:** Just showing total memory , not used/available memory following this [comment](https://github.com/ceph/ceph/pull/42600#issuecomment-892011569)
> 
> **Raw Capacity:** `hdd_capacity + flash_capacity ` . I'm not much familiar with `flash_capacity` field in gather_facts (does it includes `SSD` & `NVMes`) , @pcuzner maybe you familiar with it?
> 
> **Disk Types:** disk types with their count. How to show counts for SSD/NVME , maybe to use `flash_count` field from gather_facts?
> 
> **NIC:** I think in the design it represents the bandwidth of a Ethernet interface maybe, but I don't see any corresponding info in gather_facts, it just exposes the `nic_count`.
> 
> I'm attaching the [payload](https://pastebin.com/rZq421vt) of gather_facts too, in case if any other relevant fields needs to be exposed in host list.

@avanthakkar when deciding what information display and how to, it's useful to reflect on why this information is meaningful for the user and if there are better ways of using it than simply "displaying it":

* Things like CPU count or Memory avail. are mostly useful to guarantee [Ceph resource recommendations for deploying OSDs and other services](https://docs.ceph.com/en/latest/start/hardware-recommendations/). For example: 1-2 CPU core/OSD, 2-4 CPU cores/MDS, 4-8 GB RAM/OSD...
  * Not sure if already planned, but probably Cephadm should enforce these policies and not deploy service if the required resource capacity is not met.
  * New Prom Alerts could also be fired up when some of these requirements are not met (although OSD service allocation might be stateful and not prone to change often).
* On the other hand: number of disks, disk types and total capacity, are key to decide on which hosts to deploy OSDs.
* Same for network bandwidth (rather than just NICs... what's important here is the bandwidth of each NIC: 1, 10, 25, 40, 100 Gbit... ).

One suggestion: why not sharing the screenshot with the ceph-users mailing list requesting help from the community? We can ask them "what kind of information you would find useful when adding hosts to the cluster".

pcuzner commented:
@avanthakkar it's nice to see this moving forward. The original tracker for the Ui integration was https://tracker.ceph.com/issues/47220 Any reason for the design here not reflecting that proposal?

pcuzner reviewed (dismissed):
@avanthakkar To answer your questions; the cpu and ram totals are the values to show - since these reflect the spec of the server

I agree with Ernesto about the services, but take your point. The issue for me is that the column heading is services, but the content is the daemons so something needs to change!

Version makes no sense. A host can hold multiple ceph versions and prometheus, alertmgr etc so I'd remove. it's not a facet of the host anyway. 

flash capacity is a total of all non-rotational media - so ssd and nvme 

gather-facts does provide all NIC information in the interfaces section. nic_count isn't the only thing you get!

Using the OCS mockups for inspiration is a great idea - but please remember that for OCS, OCP owns the hardware so the OCS components don;t have tp do much. For Ceph, we have the direct relationship with the platform, so showing more detailed information would be beneficial.

I would hope that we can show an overview row of the host, and then expand the host to show the details for the host, like in the mockup https://tracker.ceph.com/issues/47220

alfonsomthd commented:
> I was wondering how to handle `api/host?facts=true` case if there's no orchestrator present. 
As you retrieving hosts info and adding "gather facts" info, we shouldn't throw any error, but make user clear that the afoementioned info is not available. `Nan` or `undefined` seems a bug XD, why not shoing sth like "Not Avail." ?

alfonsomthd commented on src/pybind/mgr/dashboard/controllers/host.py:
```diff
@@ -272,15 +290,14 @@ class Host(RESTController):
-        return get_hosts(from_ceph, from_orchestrator)
+    def list(self, sources=None, facts=False):
+        hosts = get_hosts(sources)
+        if facts == 'true':
```
```suggestion
        if str_to_bool(facts):
```

alfonsomthd commented on src/pybind/mgr/dashboard/controllers/host.py:
```diff
@@ -153,31 +153,49 @@ def merge_hosts_by_hostname(ceph_hosts, orch_hosts):
     """
+    from_ceph = True
+    from_orchestrator = True
+    if sources is not None:
```
```suggestion
    if sources:
```

alfonsomthd commented on qa/tasks/mgr/dashboard/test_host.py:
```diff
@@ -79,6 +79,27 @@ def test_host_list_with_sources(self):
+        resp_hostnames = {host['hostname'] for host in data}
+        self.assertEqual(test_hostnames, resp_hostnames)
+        for server in data:
+            self.assertIn('cpu_count', server)
```
You can create an expected gather-facts info dictionary and use intersection to check if the info is there.

alfonsomthd commented on qa/tasks/mgr/dashboard/test_host.py:
```diff
@@ -79,6 +79,27 @@ def test_host_list_with_sources(self):
+        data = self._get('{}?facts=false'.format(self.URL_HOST))
+        self.assertStatus(200)
+        for server in data:
+            self.assertNotIn('cpu_count', server)
```
Same as above.

avanthakkar commented:
> > I was wondering how to handle `api/host?facts=true` case if there's no orchestrator present.
> 
> As we're retrieving hosts info always and as a plus adding "gather facts" info, we shouldn't throw any error, but make clear to the user that the aforementioned info is not available. `Nan` or `undefined` seems a bug XD, why not showing something like "Not Avail." ?

Yes makes sense. But what abt endpoint bcz facts can't be fetched if no orch, for now I'm returning data w/o facts but not sure if that's the correct approach as u requesting `?facts=true` and returning dtaa which doesn't contain it? that's why I was considering the option of throwing error , but yeah tht cause an issue with frontend then.

alfonsomthd commented:
> Yes makes sense. But what abt endpoint bcz facts can't be fetched if no orch, for now I'm returning data w/o facts but not sure

Let's recap:
Is gather-facts info MANDATORY for Dashboard correct behavior? If yes, then throw the appropriate error (e.g., 404)
What Dashboard should do in the scenario that Orch. communication fails? It should show an error or try to show the collected data and warn the user that the communication with orch. is not available?

avanthakkar commented:
> > Yes makes sense. But what abt endpoint bcz facts can't be fetched if no orch, for now I'm returning data w/o facts but not sure
> 
> Let's recap:
> Is gather-facts info MANDATORY for Dashboard correct behavior? If yes, then throw the appropriate error (e.g., 404)
 
> What Dashboard should do in the scenario that Orch. communication fails? It should show an error or try to show the collected data and warn the user that the communication with orch. is not available?
 
 Only if Orch. is present. 

I think we should still show the other details available in frontend (maybe we can subscribe to `facts?=false` in case if orch.available=false ?). But for the endpoint I think we should throw error, as we can't return a 200 response with `?facts=true` which actually doesn't hv any `gather_facts` present. 
**To summarize my proposal:** 
Orch. = Yes: frontend-> subscribes(facts?=true) , all happy.
Orch. = No: frontend-> subscribes(facts?=false) and hides the gahter_facts columns or shows not available, backend-> for ?facts=true throws error

alfonsomthd commented on src/pybind/mgr/dashboard/controllers/__init__.py:
```diff
@@ -1105,3 +1105,11 @@ def validate_args(*args, **kwargs):
         return validate_args
     return decorator
+
+def merge_list_of_dicts_by_key(target_list: list, source_list: list, key: str):
```
This should be placed in `src/pybind/mgr/dashboard/tools.py`.

Can you add tests for this method to ensure it works properly?

avanthakkar commented:
> @avanthakkar To answer your questions; the cpu and ram totals are the values to show - since these reflect the spec of the server
> 
> I agree with Ernesto about the services, but take your point. The issue for me is that the column heading is services, but the content is the daemons so something needs to change!
> 
Yes but for now to not break the perf_counters link , how abt I just rename the column name to `Daemons`? @alfonsomthd @nizamial09 
> Version makes no sense. A host can hold multiple ceph versions and prometheus, alertmgr etc so I'd remove. it's not a facet of the host anyway.
> 
> flash capacity is a total of all non-rotational media - so ssd and nvme
> 
> gather-facts does provide all NIC information in the interfaces section. nic_count isn't the only thing you get!
> 
Yes, I'm exposing the link speed of Ethernet interface (in GbE) in NIC column now. So wht u think abt this? I'm also planning to expose more info in tooltip in case if there are multiple Eth. interfaces to show their speeds too as a follow-up.
> Using the OCS mockups for inspiration is a great idea - but please remember that for OCS, OCP owns the hardware so the OCS components don;t have tp do much. For Ceph, we have the direct relationship with the platform, so showing more detailed information would be beneficial.
> 
> I would hope that we can show an overview row of the host, and then expand the host to show the details for the host, like in the mockup https://tracker.ceph.com/issues/47220

avanthakkar commented:
jenkins test make check

avanthakkar commented:
jenkins test dashboard cephadm

aaSharma14 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.html:
```diff
@@ -67,4 +67,10 @@
+<ng-template #orchTmpl>
+  <span i18n
+        i18n-ngbTooltip
+        ngbTooltip="Data will be available only if Orchestrator is available.">Not Avail.</span>
```
IMO we should avoid using abbreviations here. Can we change it to `Unavailable` or `No Data`.

avanthakkar commented:
jenkins test make check

alfonsomthd commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:
```diff
@@ -347,24 +409,27 @@ export class HostsComponent extends ListWithDetails implements OnInit {
-            const permission = this.permissions[typeToPermissionKey[service.type]];
-            service.canRead = permission ? permission.read : false;
-            return service;
+    this.sub = this.timerService
```
As pero ur previous conversation you need to refactor this as currently it's creating a new subscription for each call.
Also probably you don't need the timerService.

avanthakkar commented:
jenkins test dashboard cephadm

sebastian-philipp commented on src/pybind/mgr/dashboard/tools.py:
```diff
@@ -828,3 +828,13 @@ def find_object_in_list(key, value, iterable):
     return None
+
+
+def merge_list_of_dicts_by_key(target_list: list, source_list: list, key: str):
+    target_list = {d[key]: d for d in target_list}
+    for sdict in source_list:
+        if bool(sdict):
+            if sdict[key] in target_list:
+                target_list[sdict[key]].update(sdict)
+    target_list = [value for value in target_list.values()]
+    return target_list
```
Note we have a similar function already: 

https://github.com/ceph/ceph/blob/f938712b7c81f9e004827d1dc2fe0f005ee1b3b0/src/pybind/mgr/mgr_util.py#L402

avanthakkar replied:
But this merging is based on a key and it merges list[dict]. Maybe should I add this to mgr_util?

sebastian-philipp replied:
you've re-implemented merge_dicts in your merge_list_of_dicts_by_key funciton

avanthakkar replied:
Not exactly, I can't directly consume tht func , I still need some modification after tht ofr e.g. to return all as a list and tht wld require another method, so I implemented all in 1 method, but I can try using `merge_dicts` in my current method.

avanthakkar commented on src/pybind/mgr/dashboard/frontend/cypress/integration/page-helper.po.ts:
```diff
@@ -206,7 +206,7 @@ export abstract class PageHelper {
     this.waitDataTableToLoad();
 
-    return cy.get('.datatable-header').its(index).find('.datatable-header-cell-label');
+    return cy.get('.datatable-header').its(index).find('.datatable-header-cell');
```
removed `-label` to also include headers of type cell-template

sebastian-philipp commented on src/pybind/mgr/dashboard/controllers/host.py:
```diff
@@ -272,15 +278,23 @@ class Host(RESTController):
-        from_ceph = 'ceph' in _sources
-        from_orchestrator = 'orchestrator' in _sources
-        return get_hosts(from_ceph, from_orchestrator)
+    @RESTController.MethodMap(version='1.1')
```
Sorry for my ignorance here: What happens to the existing 1.0 implementation? cc @epuertat

avanthakkar replied:
We already hv [API-versioning](https://github.com/ceph/ceph/pull/35769) implemented, so any changes in endpoint will require some version change, in this case as endpoint provides extra info now and it's stable so 1.1. If you try API call passing 1.0 version in header it'll thow version mismatch error.

avanthakkar replied:
So in case if this is merged & backported to next pacific release 1.0 won't work for this endpoint.

sebastian-philipp replied:
Which mean, we're breaking the API for no reason at all, right? You didn't need to change any tests in `qa/tasks/mgr/dashboard/test_host.py`, cause you didn't add any backward incompatible changes.

sebastian-philipp replied:
I mean how are client outside of our git tree supposed to cope with it? Are they supposed to try the 1.1 API and if that doesn't exist, they should use the 1.0 API? That's creating a ton of unnecessary work for everyone writing API clients

avanthakkar replied:
We update [openapi spec](https://github.com/ceph/ceph/blob/0da1fe3180d6f6d05ed997b19190998977938e41/src/pybind/mgr/dashboard/openapi.yaml#L3214) with the new versioned header too, so as long as u have updated spec I think it won't be any issue to generate API client libraries,etc.
https://github.com/OpenAPITools/openapi-generator
https://swagger.io/tools/swagger-codegen/

avanthakkar replied:
f> Which mean, we're breaking the API for no reason at all, right? You didn't need to change any tests in `qa/tasks/mgr/dashboard/test_host.py`, cause you didn't add any backward incompatible changes.

The endpoint is still backward compatible, as even if u don't pass the new parameter exposed (`facts`) it still returns the same payload as 1.0 , but it only needs to be passed the correct version. But ok I think now I got your point that <=1.1 should work too in this case 1.0, sorry maybe I got confused initially. So will need some versioning conditions to be handled here.

sebastian-philipp replied:
The idea behind API versioning is to **avoid** breaking existing clients. If this PR is backported to pacific, we're actually breaking existing clients cause clients now need to get the 1.1 version.

avanthakkar replied:
Yes I agree. I've created a new tracker for this https://tracker.ceph.com/issues/52480 ,so will fix it ASAP.

epuertat replied:
@sebastian-philipp as we're merging this to a feature branch, we are giving these more leeway, but you're right: this will break support from legacy API consumers.

Even additive changes to the API [require minor version increases](https://tracker.ceph.com/issues/40909#Versioning-semantics-and-practices), since the main goal of the versioning is to gracefully detect cross-version issues, rather that letting exceptions fly around.

For the methods `a(x)` (1.0) and `a(x, y=default)` (1.1) we need to cover the following cross-version situations:
* 1.0 client calling `a(x_0)` on the 1.1 method `a(x,y)`: :heavy_check_mark: (it'll work due to the way Python deals with optional args)
* 1.1 client calling `a(x_0, y_0)` on the 1.0 method `a(x)`: :x:

Our versioning commitment is that a client consuming a version X.Y will be able to run it against any server providing version X.Z where Y <= Z (minor version/backward support).

sebastian-philipp replied:
> For the methods `a(x)` (1.0) and `a(x, y=default)` (1.1) we need to cover the following cross-version situations:

Instead of python functions, I'm more concerned about clients directly accessing the API.

> Our versioning commitment is that a client consuming a version X.Y will be able to run it against any server providing version X.Z where Y <= Z (minor version/backward support).

Thank you.

avanthakkar commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:
```diff
@@ -333,6 +367,30 @@ export class HostsComponent extends ListWithDetails implements OnInit {
+            ethLinkSpeed = hostInterface['speed'];
+          }
+        });
+        hostKey['ethLinkSpeed'] = `${ethLinkSpeed} GbE`;
```
Gather facts are already introduced in cephadm [here](https://github.com/ceph/ceph/pull/41816) . This transformation is just to adjust units for some fields to be used in dashboard else it's using the backend info only  like cpu_count, hdd_count, memory_total_kb, etc..

nizamial09 commented:
dashboard cephadm fails will be fixed when https://github.com/ceph/ceph/pull/42919 is merged.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.html:
```diff
@@ -10,6 +10,7 @@
                 [columns]="columns"
                 columnMode="flex"
                 (fetchData)="getHosts($event)"
+                [toolHeader]="!hideToolHeaders"
```
```suggestion
                [toolHeader]="!clusterCreation"

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:
```diff
@@ -40,15 +42,23 @@ const BASE_URL = 'hosts';
+  flashTmpl: TemplateRef<any>;
   @Input()
   clusterCreation = false;
+  @Input()
+  hideToolHeaders = false;
```
Do we need this? We can use the `clusterCreation` flag instead of `hideToolHeaders`

avanthakkar replied:
Yes bcz for review section we don't need the toolHeader but for add-host section we do..so can't just rely on single variable here..need a new one.

nizamial09 replied:
Understood!

aaSharma14 reviewed (approved):
Looks Great @avanthakkar

pereman2 reviewed (approved):
looking fine

aaSharma14 commented:
jenkins test dashboard

aaSharma14 commented:
jenkins test dashboard cephadm

epuertat commented on src/pybind/mgr/dashboard/controllers/host.py:
```diff
@@ -272,15 +278,23 @@ class Host(RESTController):
+        orch = OrchClient.instance()
+        if str_to_bool(facts):
+            if orch.available():
+                hosts_facts = orch.hosts.get_facts()
+                return merge_list_of_dicts_by_key(hosts, hosts_facts, 'hostname')
```
The fact that we need this `merge_list_of_dicts_by_key()` weird function is a good smell of something not very healthy. I think we shouldn't make our back-end methods just by-passers of data. We should always gather the data we need from the inner/deeper layers, compose our response with our own format and only return the minimal set of required info, and that alone...

I know this has not started with this PR, it's a long tradition of letting data pass unprocessed through our back-end (which IMHO has always been a bad practice, as we don't have explicit data models of the info we deal with).

So, how can we commit to version the API if we're the owners of the data model we're returning... We'll always be bound to whatever changes other components introduce.

Why not creating an explicit model of the data you're returning here? We're actually validating that implicit model in the API tests... but surprisingly that model is not in our production code, only in test code... (I think these are partly the perils of type-less languages).

avanthakkar replied:
Yes, as discussed I'm trying to  design model for `HostList`, but while doing it there are many cleanups I have to do first, so I was thinking how about I deal with all these cleanups [here](https://github.com/ceph/ceph/pull/41959) itself and move forward with this PR without introducing any more changes?

epuertat replied:
Can I help you with the model here?

This could be the model:
```py
class HostFacts(NamedTuple):
  IPv4_addresses: List[IPv4Address]
  IPv6_addresses: List[IPv6Address]
  ...
  memory_total_bytes: int
  hdd_capacity_bytes:int
  flash_capacity_bytes: int
  raw_capacity: int

  
class Host(NamedTuple):
  name: str
  labels: List[str]
  ...
  facts: HostFacts
```

What we want here is something like this:
```suggestion
                return [Host(hostname=host.hostname, ...) for host in get_hosts()]
```

But we have several challenges here:
* Instead of manually entering all the class init arguments, we could use the `Host(**host)` construct and filter out the arguments our class doesn't support. The `NamedTuple` doesn´t allow to override the __init__ method, but we can create a factory method (@classmethod):
```py
  @classmethod
  from_dict(cls, host, facts):
    return cls(facts=facts, **{key: value for key, value in host.items() if key in cls._fields})
```
* If we need to change field names, I'd rather create an external helper for renaming keys in dictionaries.
* Additionally, I'd check what model (if any) the `get_hosts()` uses and use that if possible. If we do this, we'll discover a `HostSpec` class in `src/python-common/ceph/deployment/`. We'll also find a `HostFacts` class under `src/pybind/mgr/cephadm/configchecks.py`.
* Another issue is that we need to populate data from 2 different sources `get_hosts` and `get_facts` and those are 2 lists, not dicts. You have used this `merge_list_of_dicts_by_key` and directly return what you get from there... I'd rather create an index (or dict) for the facts and that's all.

This would look like this:
```py
  facts_index = {facts['hostname']: fact for fact in facts}
  return [ost.from_dict(host, facts_index[host.hostname]) for host in get_hosts()]
```

epuertat reviewed (changes requested):
My main suggestion is that we merge the code to deal with minor API versions prior to this PR, so you don't need to add that here. I'll be happy to help with the multimethod stuff in the controllers init (probably the toughest piece of changes needed).

Apart from that, there are other suggestions over there.

avanthakkar commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts:
```diff
@@ -64,17 +51,20 @@ export class CreateClusterReviewComponent implements OnInit {
       _.forEach(this.hosts, (hostKey) => {
         const labels = hostKey['labels'];
+        this.totalCPUs = this.totalCPUs + hostKey['cpu_count'];
+        this.totalMemory = this.totalMemory + hostKey['memory_total_kb'] * 125;
```
Converting to bytes(B)

avanthakkar replied:
will change name to `totalMemoryBytes` to make it more clear.

epuertat replied:
Oh, you're assuming that this `_kb` there stands for kilobits? No, that's probably not formally correct (it should be uppercase `B` and `Ki` for the kibi prefix), but storage/memory is almost never measured in bits (only serial link speeds/bandwidths) and you can assume that `k` means 99% of the time `Kibi` (x1024) not strictly `kilo` (x1000).

In case of doubt you may check where this data was gathered from but I'm mostly sure that this is in 1024 x byte units.

avanthakkar replied:
I assumed here tht from backend kb is kilobits so I'm converting it to KiB in UI. I can't verify if it's bits/bytes in backend.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.scss:
```diff
@@ -0,0 +1,5 @@
+cd-hosts {
+  ::ng-deep .nav {
```
I searched for this and I think ng-deep is still our best hope. Although I found one interesting solution here: https://dev.to/jwp/angular-10-ng-deep-and-host-pseudo-classes-1kbg, didn't tried it out yet. But according to most, a replacement for ng-deep is still a work in progress.

avanthakkar commented on qa/tasks/mgr/dashboard/test_auth.py:
```diff
@@ -154,7 +154,7 @@ def test_logout(self):
             "redirect_url": "#/login"
         })
-        self._get("/api/host")
+        self._get("/api/host", version='1.1')
```
> My main suggestion is that we merge the code to deal with minor API versions prior to this PR, so you don't need to add that here. I'll be happy to help with the multimethod stuff in the controllers init (probably the toughest piece of changes needed).
> 
I've handled the minor version requests for now, for multimethod or other changes required to handle mutli versions for endpoint with backward compatibility I'm thinking on moving tht to seperate PR if that's ok, tracker already created https://tracker.ceph.com/issues/52480.  It's a very interesting problem to solve though!
> Apart from that, there are other suggestions over there.

avanthakkar replied:
Done!

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/hosts/hosts.component.ts:
```diff
@@ -336,6 +378,20 @@ export class HostsComponent extends ListWithDetails implements OnInit {
+        hostKey['raw_capacity'] = hostKey['hdd_capacity_bytes'] + hostKey['flash_capacity_bytes'];
+      });
+    } else {
+      for (let column = 5; column < this.columns.length; column++) {
+        this.columns[column]['prop'] = '';
+        this.columns[column]['cellTemplate'] = this.orchTmpl;
+      }
```
What's this for?

avanthakkar replied:
It's for  showing `Unavailable` for gather facts columns when orch is not available.

avanthakkar commented on src/pybind/mgr/dashboard/controllers/__init__.py:
```diff
@@ -713,7 +713,8 @@ def inner(*args, **kwargs):
                     raise cherrypy.HTTPError(415, "Unable to find version in request header")
 
-                if req_version and req_version == version:
+                if req_version and (req_version <= version
```
Done, thanks.

avanthakkar commented:
I see a long loading time issue for  gather_facts, taking minutes (4-20min depending on hardware resources) . I think it's not just for gather_facts but for physical disks as well, as mentioned [here](https://github.com/ceph/ceph/pull/42583#discussion_r695759635) , so my question is how to improve that loading time? Does this PR https://github.com/ceph/ceph/pull/43057 takes care of it or  anything  more can be done for improving loading time for gather_facts? @sebastian-philipp @Daniel-Pivonka @jmolmo @adk3798

epuertat reviewed (changes requested):
Thanks for addressing my suggestions @avanthakkar !

However, I may assume that we finally don't go with a back-end model, but the front-end model has been a must so far (I'd personally trade off the back-end one for the front-end, if we can only afford one). If we chose not improve our code with back-end validation for agility reasons ok, but at least don't lower the standard we have previously set.

liewegas commented:
> I see a long loading time issue for gather_facts, taking minutes (4-20min depending on hardware resources) . I think it's not just for gather_facts but for physical disks as well, as mentioned [here](https://github.com/ceph/ceph/pull/42583#discussion_r695759635) , so my question is how to improve that loading time? Does this PR #43057 takes care of it or anything more can be done for improving loading time for gather_facts? @sebastian-philipp @Daniel-Pivonka @jmolmo @adk3798

Looking at the code, cephadm appears to do the "right" thing: we only refresh info (daemons, devices, facts) for a host if the cached value is too old, or if it is missing.  That means simply adding a new host will trigger an immediate "refresh" of just that host, which should be quite quick.

Perhaps the dashboard is passing '--refresh' to 'orch ps' or something? That will trigger a refresh of all hosts...

avanthakkar commented:
@epuertat I've rebased the PR and removed the changes of versioning as well. So I think if the all jenkins job passes it's ready-to-merge. For the backend model I've created a [gist](https://gist.github.com/avanthakkar/5c4dbf4619ea3c80c185afcfefc28b79) for the changes I made yesterday, which can be later continued I think in this [cleanup](https://github.com/ceph/ceph/pull/41959) PR for the master.

avanthakkar commented:
jenkins test dashboard cephadm

avanthakkar commented:
jenkins test dashboard

avanthakkar commented:
jenkins test dashboard

nizamial09 commented:
@avanthakkar dashboard e2e won't pass because it is not up-to-date with master. When I rebase this feature branch against master it'll starts to pass. I am thinking of doing that on monday.

avanthakkar commented:
> @avanthakkar dashboard e2e won't pass because it is not up-to-date with master. When I rebase this feature branch against master it'll starts to pass. I am thinking of doing that on monday.

Oh okay..well in that case I think we should merge this PR first and then rebase feature branch with master.

avanthakkar commented on qa/tasks/mgr/dashboard/test_auth.py:
```diff
@@ -169,7 +169,7 @@ def test_logout(self):
             "redirect_url": "#/login"
         })
-        self._get("/api/host", set_cookies=True)
+        self._get("/api/host", set_cookies=True, version='1.1')
```
We can't remove this until this changes are in master, so for now I've to keep this in-order to pass this tests.

avanthakkar commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/create-cluster/create-cluster-review.component.ts:
```diff
@@ -99,10 +85,15 @@ export class CreateClusterReviewComponent implements OnInit {
+      _.forEach(this.hosts, (hostKey) => {
+        this.totalCPUs = this.totalCPUs + hostKey['cpu_count'];
+        // convert to bytes
+        this.totalMemory = this.totalMemory + hostKey['memory_total_kb'] / 1000;
```
Oh I forgot to add here. I think actually it should be 1000 not 1024 as it's [kiloBytes](https://en.wikipedia.org/wiki/Kilobyte)

avanthakkar commented:
jenkins test signed

epuertat commented:
Ok, I manually checked and the commit has the signed-off stuff.
